app/grandchallenge/reader_studies/views.py

```diff
@@ -1429,3 +1468,326 @@ def form_valid(self, form):
+            if interface.is_image_kind:
+                entry["value"] = [entry["value"]]
+            form = DisplaySetInterfacesCreateForm(
+                data=entry,
+                pk=None,
+                interface=interface.pk,
+                user=self.request.user,
+                reader_study=self.reader_study,
+                auto_id="%s",
+            )
+            if form.is_valid():
+                cleaned = form.cleaned_data
+                validated_data[cleaned["interface"].slug] = cleaned["value"]
+            else:
+                errors.update({entry["interface"]: form.errors["value"]})
+        if errors:
+            raise ValidationError(errors)
+        return validated_data
+
+    def create_display_set(self, data):
+        ds = DisplaySet.objects.create(reader_study=self.reader_study)
+        ds.order = data.pop("order")
+        ds.save()
+        for slug in data:
+            if not data[slug]:
+                # Field is not filled in the form
+                continue
+            interface = ComponentInterface.objects.get(slug=slug)
+            civ = ComponentInterfaceValue(interface=interface)
+            if interface.requires_file:
+                user_upload = data[slug]
+                transaction.on_commit(
+                    partial(
+                        add_file_to_display_set.apply_async,
+                        kwargs={
+                            "user_upload_pk": str(user_upload.pk),
+                            "interface_pk": str(interface.pk),
+                            "display_set_pk": str(ds.pk),
+                        },
+                    )
+                )
+            elif interface.is_json_kind:
+                civ = ComponentInterfaceValue.objects.create(
+                    value=data[slug], interface=interface
+                )
+                ds.values.add(civ)
+            elif interface.is_image_kind:
+                us = RawImageUploadSession.objects.create(
+                    creator=self.request.user,
+                )
+                us.user_uploads.set(data[slug])
+                us.process_images(
+                    linked_task=add_image_to_display_set.signature(
+                        kwargs={
+                            "display_set_pk": str(ds.pk),
+                            "interface_pk": str(interface.pk),
+                        },
+                        immutable=True,
+                    )
+                )
+
+    def return_errors(self, errors):
+        return JsonResponse(errors, status=400)
+
+    def form_invalid(self, form):
+        errors = form.errors
+        try:
+            self._process_new_interfaces()
+        except ValidationError as e:
+            errors.update(e.message_dict)
+        return self.return_errors(errors)
+
+    def form_valid(self, form):
+        try:
+            data = self._process_new_interfaces()
+        except ValidationError as e:
```

Comment:
Why is the validation being done in the view and not in the form? Validation errors should not occur here.

Reply:
The validation is done un the form, but this view combines multiple forms and collects the errors from each of them.